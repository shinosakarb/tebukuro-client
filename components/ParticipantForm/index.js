// @flow
import React, { Component } from 'react'
import ParticipantButton from '../buttons/ParticipantButton'
import TextInput from '../forms/TextInput'

type Props = {
  onSubmit: Function,
  hasEventWaitlist: boolean,
  eventId: ?number,
  errors: ?string[],
  message: string,
}
type State = {
  name: string,
  eventId: ?number,
}

export default class ParticipantForm extends Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      name: '',
      eventId: props.eventId,
    }
  }

  onChangeHandler = (e: SyntheticInputEvent<>) => {
    this.setState({
      ...this.state,
      [e.target.id]: e.target.value,
    })
  }

  onSubmitHandler = (e: SyntheticEvent<>) => {
    e.preventDefault()
    this.props.onSubmit(this.state)
  }

  render() {
    return (
      <div>
        <h3>Event participation form</h3>
        { this.props.errors &&
          <ul>
            { this.props.errors.map(error =>
              <li>{ error }</li>)}
          </ul>
        }
        { this.props.message }
        <form onSubmit={this.onSubmitHandler}>
          <TextInput id="name" value={this.state.name} onChange={this.onChangeHandler} />
          <ParticipantButton hasEventWaitlist={this.props.hasEventWaitlist} />
        </form>
      </div>
    )
  }
}
