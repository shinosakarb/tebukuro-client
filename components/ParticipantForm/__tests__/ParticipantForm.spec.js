import React from 'react'
import { mount, shallow } from 'enzyme'
import shallowToJson from 'enzyme-to-json'
import ParticipantForm from '../index'

const inputValues = {
  name: 'participant1',
}

const testProps = {
  eventId: 1,
  participateButtonText: '参加登録',
}

const errorProps = {
  ...testProps,
  errors: ['名前を入力して下さい'],
}

describe('ParticipantForm', () => {
  it('renders participant form component', () => {
    const wrapper = shallow(<ParticipantForm {...testProps} onSubmit={jest.fn()} />)
    const tree = shallowToJson(wrapper)

    expect(tree).toMatchSnapshot()
  })

  it('renders participant form component with error message', () => {
    const wrapper = shallow(<ParticipantForm {...errorProps} onSubmit={jest.fn()} />)
    const tree = shallowToJson(wrapper)

    expect(tree).toMatchSnapshot()
  })

  describe('when submit', () => {
    describe('with correct argument', () => {
      const onSubmit = jest.fn()
      const wrapper = mount(<ParticipantForm {...testProps} onSubmit={onSubmit} />)

      const nameElement = wrapper.find('[type="text"]')

      nameElement.simulate(
        'change',
        { target: { id: 'name', value: inputValues.name } },
      )
      wrapper.find('[type="submit"]').simulate('submit')

      it('should call onSubmit once.', () => {
        expect(onSubmit).toHaveBeenCalledTimes(1)
      })

      it('should call onSubmit with correct argument.', () => {
        expect(onSubmit).toBeCalledWith({ ...inputValues, eventId: testProps.eventId })
      })
    })
  })
})
