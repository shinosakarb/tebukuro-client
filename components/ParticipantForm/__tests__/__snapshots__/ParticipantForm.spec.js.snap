// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParticipantForm renders participant form component 1`] = `
<div>
  <h3>
    Event participation form
  </h3>
  <form
    onSubmit={[Function]}
  >
    <div>
      <label
        htmlFor="name"
      >
        name
        <input
          id="name"
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
    </div>
    <ParticipantButton />
  </form>
</div>
`;

exports[`ParticipantForm renders participant form component with error message 1`] = `
<div>
  <h3>
    Event participation form
  </h3>
  <ul>
    <li>
      名前を入力して下さい
    </li>
  </ul>
  <form
    onSubmit={[Function]}
  >
    <div>
      <label
        htmlFor="name"
      >
        name
        <input
          id="name"
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
    </div>
    <ParticipantButton />
  </form>
</div>
`;
